File: ContainerWithMostWater.cs
using System;

//from https://leetcode.com/problems/container-with-most-water/
public class Solution {
    
    public int MaxArea(int[] height) {
        int bestResult = -1;
        
        int l = 0; 
        int r = height.Length-1;
        
        //start from the edges and step one closer to each other each loop until you have the best results
        
        //this way, the width between the two points will automatically be reduced
        //by one each loop and it'll ensure the higher one is preserved
        while(l < r)
        {
            int currentComparison = CalculateArea(height[l], l, height[r], r);
            if(currentComparison > bestResult)
            {
                bestResult = currentComparison;
            }
            
            if(height[l] < height[r])
            {
                l++;
            }
            else
            {
                r--;
            }
        }
        
        return bestResult;
    }
    
    public int CalculateArea(int leftHeight, int leftIndex, int rightHeight, int rightIndex)
    {
        /*int height = Math.Max(leftHeight, rightHeight);
        int width = rightIndex - leftIndex;
        return (height*width);*/
        
        return (Math.Min(leftHeight, rightHeight)) * (rightIndex - leftIndex);
    }
}

File: TheBlind75/containerWithMostWater.cpp
//https://leetcode.com/problems/container-with-most-water/submissions/
class Solution {
public:
    int maxArea(vector<int>& height) {
        //O(n)
		
        //iterate over the vector with two iterators
        //move the lower of the two. calculate the height
        //based on the difference between the two

        int greatest = 0;

        int l = 0; int r = height.size()-1;
        while(l < r)
        {
            long current = min(height[l], height[r]) * (r-l);
            if(current > greatest)
            {
                greatest = current;
            }

            if(height[l] < height[r])
            {
                l++;
            }
            else
            {
                r--;
            }
        }

        return greatest;
    }
};